packages/zoe/src/contracts/operaConcertTicket.js

```diff
@@ -0,0 +1,138 @@
+/* eslint-disable no-use-before-define */
+import harden from '@agoric/harden';
+import produceIssuer from '@agoric/ertp';
+import { makeZoeHelpers } from './helpers/zoeHelpers';
+
+/*
+  Roles in the arrangement:
+  - Contract creator: describes the contract with:
+    - number of seats, show, date/time of start
+    - expected (ERTP) amount per ticket (we assume all tickets cost the same)
+  - Smart Contract: 
+    - mints the tickets
+    - provides the seats
+  - Auditorium (unique contract seat, usually taken by the contract creator): the person hosting 
+  the Opera show, selling the tickets and getting the payment back
+  - Ticket buyers (contract seat created on demand): 
+    - can see the available opera show seats
+    - can consult the terms
+    - can redeem the zoe invite with the proper payment to get the ticket back
+*/
+
+export const makeContract = harden(async zoe => {
+  // Create the internal ticket mint
+  const { issuer, mint, amountMath } = produceIssuer('Opera tickets', 'set');
+  await zoe.addNewIssuer(issuer, 'Ticket');
+
+  // create Zoe helpers after zoe.addNewIssuer because of https://github.com/Agoric/agoric-sdk/issues/802
+  const { rejectOffer } = makeZoeHelpers(zoe);
+
+  const {
+    terms: { show, start, count },
+  } = zoe.getInstanceRecord();
+
+  const inviteHandleByTicketNumber = new Map();
+
+  function completeAmountKeywordRecord(amountKeywordRecord) {
+    const { issuerKeywordRecord } = zoe.getInstanceRecord();
+
+    const completed = { ...amountKeywordRecord };
+
+    for (const [keyword, keywordIssuer] of Object.entries(
+      issuerKeywordRecord,
+    )) {
+      if (!(keyword in completed)) {
+        completed[keyword] = keywordIssuer.getAmountMath().getEmpty();
+      }
+    }
+
+    return harden(completed);
+  }
+
+  const auditoriumSeat = harden({
```

Comment:
I think we can make this contract simpler by separating the creation of the ticket amounts from the actual minting of tickets. We can first create amounts for all of the tickets to be sold, and this can be what is returned from `getAvailableTickets`. That way, `getAvailableTickets` has nothing to do with offers or invites. Then, when a user chooses to purchase a ticket or a number of tickets, we can mint the tickets on demand, remove their extent from the `availableTickets` amount, create a new offer to escrow the tickets with Zoe, and then reallocate such that 1) the user's offer is allocated the tickets, 2) an offer representing the administrator is allocated the money, and 3) the new offer now is allocated nothing. This would follow the usage pattern we had in Autoswap. 

I can see why you had an administrative method to withdraw the money from the contract. To sum up the problem, it's that the seller of the tickets wants to receive a payout of money on demand, and doesn't want to wait until the tickets are sold to be able to receive the money? What if we simplify that constraint, and say that the seller of tickets won't withdraw their money until they are done selling tickets, regardless if they have all been sold or not. If we have an offer representing the entire ticket seller's position in the contract (not the short term offers created in the contract), then if that offer has the `exit` rule `onDemand`, when the ticket seller calls `cancel`, it would give the ticket seller their money and, if we write the contract to do so, shut down the contract.

Reply:
> That way, getAvailableTickets has nothing to do with offers or invites. 

(`getAvailableTickets` only manipulates offerHandles, not invites... but i realize the code is confusing because of the inviteHandle/offerHandle mix. I'll fix the wording to only use `offerHandle`)

That's what i had initially written
and then i realized that keeping track of where the non-fungible tickets are and that none has been lost because of a bug and which assets belongs to whom (in the sense of "associated with which offerHandle") is the job Zoe+ERTP have been designed for
So i feel it's very relevant to leverage that instead of re-implementing it myself

And in order for Zoe+ERTP to keep track of what's still available, they need to be made aware of all tickets initially, hence minting ahead-of-time and not on-demand

I think this might be a place where the fact that the contract deals with non-fungible assets is a relevant difference with the Autoswap pattern (which certainly mints on-demand)

Finally, when all tickets are in the hands of Zoe and i have the corresponding offerHandles, i know i can trust Zoe+ERTP to tell me what is still available and what is not


> What if we simplify that constraint, and say that the seller of tickets won't withdraw their money until they are done selling tickets, regardless if they have all been sold or not.

Yes, exactly
This is what i was saying in an earlier comment

> 2) an offer representing the administrator is allocated the money

This part is hard with the contract minting the tickets
How much does the administrator trust the smart contract to perform an offer on its behalf (what you're proposing) rather than doing the offer themself after the contract produced invites (current version)?

I feel that rather than doing a 3-way offer which does work but is complicated and may lead to errors, the Opera is better to be left with the payments and invites, `redeem` early to make Zoe aware of the tickets and deal with the `payout` whenever they want (as the payment arrives or at the end; this option actually settles the question of the payouts because the Opera and not the contract is now responsible for collecting the payouts)